### ISSUE TITLE
Javascript API in SEB to query the BEK/CK (to regain key check after CORS issue #66)

### ISSUE THREAD

**Issue by jeroenhabets** [SafeExamBrowser/seb-win-refactoring#84]
Having a Javascript API in SEB to query the BEK/CK would enable us to regain key check that we lost due to CORS issue #66

Issue #66 forced us to remove our HTTP Header based key (BEK/CK) check.
In that issue @danschlet [commented on Jan 4](https://github.com/SafeExamBrowser/seb-win-refactoring/issues/66#issuecomment-753917860):

> We will anyways introduce a new Javascript API in SEB to query the BEK/CK, instead of having to check the HTTP request headers. The newer WebKit browser engine (used in SEB for macOS/iOS) doesn't allow to change request headers anymore, and now also Chromium/Blink doesn't fully support request header modifications. So this is a dead end and needs to be replaced asap.

I've searched your website, commit log and Issues/PRs, but cannot find anything on this "new Javascript API in SEB to query the BEK/CK" @danschlet mentioned. (AFAICT Developer info also still only talks about the HTTP check)

Did I overlook it? 
Or, if it is not presently supported, can you share when you expect to add support for this JS-based alternative?

P.S. appreciate your work!

**Comment by dbuechel** [comment]
No, you're correct, there currently isn't a JavaScript API for the BEK and CK in SEB 3.x for Windows. There is a prototype implementation in iOS or macOS, as far as I know, but we currently don't have a release date for this feature.

**Comment by danschlet** [comment]
The SEB JavaScript API can be tested with the current iOS beta, which can be [requested here](https://testflight.apple.com/join/egxLmc4j) (requires iPad, iPhone or iPad Touch with iOS 11 or newer).

Below some initial information (it's still in development, so the developer documentation isn't updated yet). There was some discussion about the API in [this issue](https://github.com/SafeExamBrowser/seb-mac/pull/48#issuecomment-824324010).

- Changed method to query the Browser Exam and Config Key (BEK/CK) with the new SEB JavaScript API:
1)  `void SafeExamBrowser.security.updateKeys(function callback)` 
This function needs to be invoked first (for example in <body onload="...">). Indicate a callback function (see http://safeexambrowser.org/exams/bek_ck.html) as parameter, which will be called asynchronously by SEB after updating the variables below:
2)  `SafeExamBrowser.security.browserExamKey`
3)  `SafeExamBrowser.security.configKey`
which are the BEK and CK hashed with the URL of the page. The keys are identical to the ones send in the HTTP request header when UIWebView is used. Use this new JS API in an assessment web app with BEK/CK 
and WKWebView. Also set the key browserWindowWebView to the policy "Prefer Modern" (value 3). 
3) `SafeExamBrowser.version` (new) with format appDisplayName_iOS_versionString_buildNumber_bundleID. This variable is set regardless if updateKeys() is called.

**Comment by jeroenhabets** [comment]
@dbuechel, @danschlet Thank you very much for the detailed information! If I summarize: the Javascript API is not yet 100% final though a beta has started for iOS. 
Based on the API info you shared in [previous comment](https://github.com/SafeExamBrowser/seb-win-refactoring/issues/84#issuecomment-839954031) if/when it _were_ to make it into SEB for Windows it seems to be relatively simple to implement for our product.

As our customers exclusively use SEB for Windows. I guess they have to be patient (or perhaps offer to co-fund).

Could you keep this issue open so we can watch it, please?

**Comment by ScottDellinger** [comment]
@danschlet @dbuechel As we were going to be using the documented HTTP headers method to secure the SEB configuration, this leaves us in a bit of a situation.

The JavaScript API sounds like a great solution, but it does not sound like this has been implemented in the Windows version. Are there plans for it to be included? We'd like to avoid having to have a password on the files, as this adds an additional layer of complexity for students who just want to do their exam/activity, but of course don't want students to be able to change the config settings.

Is there a recommended method to achieve this goal?

**Comment by danschlet** [comment]
The plan is to bring the JavaScript to all platform versions of SEB, but unfortunately we‘re not in a position yet when we could release all new features simultaneously for all platforms. 

Normally new SEB versions are downward compatible, but sometimes there are exceptions when it‘s technologically not possible (or in the case of SEB 3.x vs. SEB 2.x because of a major refactoring, some features of 2.x will be added to later versions of 3.x).

This API is essential for SEB 3.0 for macOS and iOS for technological reasons: The modern WebView (WKWebView) doesn‘t allow to change HTTP headers for all requests, it‘s simply impossible due to the changed Apple APIs. At the same time, WebRTC is only supported in WKWebView, not the older WebViews. So I had to add this API in a timely matter and I had to test and modify the implementation in the real code (I couldn’t just specify it on paper months ago). Therefore it is impossible to have it at the same time in the Windows version, sorry for that.

So there are two possibilities for now:

1. Implement support for both the HTTP header check and the JavaScript API for the ConfigKey (and if required for the BrowserExamKey BEK). As mentioned, the keys will be same for the same web page (URL). So in your web application you can check for the availability of the `SafeExamBrowser.security.updateKeys()` function (or the `SafeExamBrowser.version` variable). If it‘s available, then the JavaScript API is supported and you can get the ConfigKey/BEK from it. Otherwise get the ConfigKey/BEK from the HTTP headers.
2. If you can use WebRTC in a separate browser window/tab (not the main one where the exam is displayed), then you can check the HTTP headers in the main window/tab. There is a new SEB setting (key `browserWindowWebView`, see below), and with its default value (integer value 2) the main browser window/tab will use the classic WebView, when settings key `sendBrowserExamKey = true`, but the modern WebView will be used in all additional browser windows/tabs. If you chose option 1 and support both HTTP headers and JavaScript API for the ConfigKey/BEK, then set the key `browserWindowWebView` to the policy "Prefer Modern" (integer value 3, see below for more information).


- Added support for the modern WebKit browser engine (WKWebView). WKWebView currently doesn't support URL content filters (Network > URL Filter > Filter Embedded Content), embedded server certificates and certificate pinning (Network > Certificates > Pin Certificates) and sending the Browser Exam and Config Key in HTTP headers (Exam Session > Use Browser & Config Keys). If one of these settings is enabled, the classic WebView is used in the Automatic policy. Prefer Modern ignores the 'Use Browser & Config Keys' option in new tabs on a different server or generally.
- The default setting for "Select Browser Engine Policy" (key `browserWindowWebView`), "Prefer Modern in New Tab+Different Host", ensures that you still can use testing systems/LMS like Moodle which use the Browser Exam/Config Key test for SEB connecting to the exam, but webpages opened in new tabs, which have another host than the exam system open in WKWebView.
- From iOS 14.3, WKWebView supports access to the camera/microphone via WebRTC, which is needed for web-based remote proctoring or video conferencing solutions.

**Comment by jeroenhabets** [comment]
^^ decided to create a separate ticket for my questions around the specification.

**Comment by dbuechel** [comment]
> 
> 
> ^^ decided to create a separate ticket for my questions around the specification.

I allowed myself to convert the issue into a discussion: #98

**Comment by jeroenhabets** [comment]
@dbuechel thank you for that input. I'll get back to you here after we've started and made some progress.

**Comment by dbuechel** [comment]
@jeroenhabets We changed our release schedule and will release a minor update version 3.3.2 which will already contain the JavaScript API. We hope to be able to release 3.3.2 by the end of this year, but certainly faster than the originally planed version 3.4.0.

**Comment by jeroenhabets** [comment]
@dbuechel Thanks! That is excellent news for our Safe Exam Browser customers! Our releases are packed but I will see what I can do now that I have this news!

**Comment by ScottDellinger** [comment]
I've been testing this out all day today and have found some odd behavior.

If I create a stand-alone page to test this and upload it to my server and place the startURL for that page in the seb file, this works fine. But here's where it doesn't:

- User logs in to LMS
- User goes to exam activity
- We detect they are not using SEB and provide a link to download the seb config file
- SEB config file has the exam activity URL in it
- User double-clicks the SEB config file and SEB opens
- User is not logged in, so SEB takes user to login page
- User logs in and is redirected to exam page
- Key generated by the browser does not match the key generated by the same JavaScript used in the situation where it works

BUT

If I refresh the page after I've been redirected to the exam page, the keys match. Is it possible that SEB is not using the current URL to generate the key?

**Comment by dbuechel** [comment]
Thanks for your feedback, @ScottDellinger. We had similar reports recently, e.g. #241. However, if I understand you correctly, in your case both the request headers as well as the JavaScript API calculate incorrect values?

**Comment by ScottDellinger** [comment]
> Thanks for your feedback, @ScottDellinger. We had similar reports recently, e.g. #241. However, if I understand you correctly, in your case both the request headers as well as the JavaScript API calculate incorrect values?

I'm only using the JavaScript API. With the request headers method being phased out, there was no point in implementing that method in our integration with SEB. So I can only attest to the JavaScript API working this way.

**Comment by dbuechel** [comment]
Okay, I'll see that I can have another look at it. Unfortunately, I'm at the moment quite occupied with another work item. In case it would be really easy for you to verify whether the request headers actually do match, then that would be of great help in that I could pin down and solve the issue faster.

**Comment by danschlet** [comment]
I would guess that SEB doesn't use the URL of the new page if it was loaded by a server redirect request.

@ScottDellinger where you able to solve the issue with the callback function or are you using the JS API without the callback? According to @dbuechel the configKey variable can be set without using the callback in the Windows browser engine, what unfortunately didn't work in the macOS browser engine.

**Comment by ScottDellinger** [comment]
@danschlet I moved over to the Windows version because I haven't been able to get the macOS project to compile (as we've been discussing via email) and wanted to get everything else functional. Once I can get the macOS project compiled, I can make some changes to allow it to use a null reference (or try some other ideas).

I DID notice that the Windows version didn't have a problem with the callback function (which is now above the calling code, as you mentioned in email), so I'll do some more testing on the macOS version today and get back to you.

**Comment by ScottDellinger** [comment]
@dbuechel I'd love to help you out in this one, but unfortunately, our software is client-side JavaScript-based and from what I understand it's not possible to get the current request's HTTP headers in JavaScript. I looked into implementing it on the server side in our node.js function, but no luck there, either.

Our software is remote proctoring JavaScript and is injected into the current LMS activity (such as an exam) with a script tag. The headers for the initial call to the LMS are not provided to our platform (within an AWS Lambda function with API Gateway proxy) when called by the hosting page.

**Comment by dbuechel** [comment]
@ScottDellinger Aha, I see. No worries, I'll investigate the issue as soon as I get the time, hopefully I'll be able to fix it reasonably fast so that we still can keep the release deadline for version 3.3.2 by the end of this year.

**Comment by ScottDellinger** [comment]
@dbuechel I'll see if I can set up a stand alone page that will duplicate the problem in a very simple environment so we can clearly see what's going on. Should be able to have that done today.

**Comment by ScottDellinger** [comment]
@dbuechel Okay... I've managed to set up a simple example to demonstrate the issue. It appears that the HTTP header is not affected the same way that the JavaScript API is.

I've attached a zip with 2 SEB files:

- test-noredirect.seb
This file will take you to a single page with no redirect. It shows all Config keys match.

- test-redirect.seb
This file will take you to the target page, which will check for a cookie. If the cookie doesn't exist, it will redirect you to a page with a button that reads "Redirect". Clicking this button will set the cookie and redirect you back to the target page. you can see that the HTTP headers match the expected hash, but the JavaScript API doesn't. If you then refresh that page, the keys will match again.

[test.zip](https://github.com/SafeExamBrowser/seb-win-refactoring/files/7604573/test.zip)

**Comment by dbuechel** [comment]
@ScottDellinger Awesome, thanks a lot, this is a tremendous help! I hope that I can finish my other pending tasks in about a week and take up this issue afterwards, but either way I'm confident that we can solve the problem by the end of this year for the release version 3.3.2.

**Comment by danschlet** [comment]
@ScottDellinger I was able to refactor the SEB JS API implementation in the macOS version, so it also sets the CK/BEK variables directly when loading a web page, without having to call the updateKeys() function. I had to add a custom reload method, so the variables are also correctly set after reloading a web page, but this seems to work correctly now.

I will do some more testing a release a SEB 3.2 preview version containing this update JS API later this weekend.

**Comment by dbuechel** [comment]
@danschlet Should I thus remove that callback from the API for the final release? In the Windows version, it already is empty and just executed to conform with the specification.

**Comment by danschlet** [comment]
@dbuechel Please leave it for now. It's likely someone already implemented the original method using updateKeys(), so we should not break that.

**Comment by dbuechel** [comment]
Okay, I am one step further. Thanks a ton for the example you've provided, @ScottDellinger! It helped me hugely to track down the issue effectively. In the [method where I inject the JavaScript](https://github.com/SafeExamBrowser/seb-win-refactoring/blob/master/SafeExamBrowser.Browser/Handlers/RenderProcessMessageHandler.cs#L33), it appears to be the case that the resource URL does not get updated correctly for a redirect, thus the keys first get calculated with "SEBRedirect.html" and only after reloading with the actual, correct URL "SEBTarget.html":

```
2021-11-29 16:02:15.928 [26] - WARNING: [RenderProcessMessageHandler #1] BEK: 5649623ac329299124c9a36539b6c273e8422cac23b4d3498e2a3c107daf5b00
2021-11-29 16:02:15.929 [26] - WARNING: [RenderProcessMessageHandler #1] CK: f70a386b103a4ed688ad3315edab4cbb77a37118500903ceb6bf384dc46291f7
2021-11-29 16:02:15.929 [26] - WARNING: [RenderProcessMessageHandler #1] URL (frame): 'https://scottdellinger.com/SEBRedirect.php'
2021-11-29 16:02:15.930 [26] - WARNING: [RenderProcessMessageHandler #1] URL (webBrowser): 'https://scottdellinger.com/SEBRedirect.php'
...
2021-11-29 16:02:15.931 [26] - DEBUG: [Browser Instance #1] Navigated.
2021-11-29 16:02:37.436 [01] - DEBUG: [Browser Instance #1] The user confirmed reloading the current page...
...
2021-11-29 16:02:37.655 [26] - WARNING: [RenderProcessMessageHandler #1] BEK: 03aff983d03efcb9adbae6076dc73650e73108d2a48df00718e02d71742eb961
2021-11-29 16:02:37.656 [26] - WARNING: [RenderProcessMessageHandler #1] CK: 8af361a19f07e93783181c45f14c82d40822c7cbce9549266e415e5b4270d91f
2021-11-29 16:02:37.657 [26] - WARNING: [RenderProcessMessageHandler #1] URL (frame): 'https://scottdellinger.com/SEBTarget.php'
2021-11-29 16:02:37.657 [26] - WARNING: [RenderProcessMessageHandler #1] URL (webBrowser): 'https://scottdellinger.com/SEBTarget.php'
```

It isn't quite clear to me whether on redirect, that browser handler isn't executed again and whether I thus need to use another method in addition, I'll first have to study the API further and most likely ask for help in the CefSharp forum.

**Comment by ScottDellinger** [comment]
@dbuechel I am very pleased I could be of assistance. Please don't hesitate to let me know if there's anything else I can do to help!

**Comment by danschlet** [comment]
The SEB version with the updated SEB JS API (SEB 3.2pre4) is now available here.
https://github.com/SafeExamBrowser/seb-mac/releases/tag/3.2pre4

**Comment by ScottDellinger** [comment]
@danschlet I've attempted to test the new version on macOS and I'm now getting "Can't find variable: SafeExamBrowser" when I call SafeExamBrowser.security.configKey. Am I calling it incorrectly?

**Comment by danschlet** [comment]
Last time I tested it, it worked with this test page:

http://safeexambrowser.org/exams/bek_ck_new.html

I didn't test the preview version with different macOS versions yet, only with macOS 12 Monterey until now.

Generally the same SEB settings must be used as mentioned here to make sure the modern WebView is active: https://safeexambrowser.org/developer/seb-config-key.html

**Comment by ScottDellinger** [comment]
@danschlet I tested it using the files here: https://github.com/SafeExamBrowser/seb-win-refactoring/issues/84#issuecomment-979372009 without any luck. Also in Monterey.

**Comment by danschlet** [comment]
@ScottDellinger if read the instructions on the page mentioned above (I copied the relevant section below) and add the setting `browserWindowWebView=3`, then it works. I modified your two files accordingly (and changed some settings to allow to do screen shots on Mac and run on virtual machines), see attachment. 

As you can see on the attached screen shots, the configKeys match in the macOS/Windows version without redirect and in the Mac version the CK is also correct with redirect. 

As the above mentioned documentation explains, in SEB for macOS you cannot have the CK/BEK in the http request headers and in the JS API (that's why you don't see the request header values in the Mac screen shots).

```
    <key>browserWindowWebView</key>
    <integer>3</integer>

```

**How to transition to the SEB JavaScript API**

In the SEB configuration generated by Moodle, set the SEB settings key `browserWindowWebView` to the policy "Prefer Modern" (**integer value 3**), which will force the use of the modern WebView and the Javascript API in SEB versions where it is available.

Without redirect:
<img width="1001" alt="Screenshot 2021-11-30 at 01 33 41" src="https://user-images.githubusercontent.com/2464749/143965372-0ee10322-a8dc-4098-a899-bdb8246f90f2.png">
<img width="1035" alt="Screenshot 2021-11-30 at 01 37 20" src="https://user-images.githubusercontent.com/2464749/143965402-f9435fd1-f607-4e57-87e6-d73c03e54178.png">

With redirect:
<img width="1015" alt="Screenshot 2021-11-30 at 01 34 39" src="https://user-images.githubusercontent.com/2464749/143965413-3037d8fc-e1d7-4320-8c21-c7f7b35adeaf.png">
<img width="1040" alt="Screenshot 2021-11-30 at 01 38 17" src="https://user-images.githubusercontent.com/2464749/143965472-c945b79c-dac9-44a7-88dd-118e34af2381.png">


[SEB-JS-API_updated_for_macOS.zip](https://github.com/SafeExamBrowser/seb-win-refactoring/files/7622063/SEB-JS-API_updated_for_macOS.zip)

**Comment by ScottDellinger** [comment]
@danschlet Aha! Thanks for the clarification! Much appreciated!

**Comment by dbuechel** [comment]
FYI: I have asked in the CefSharp community if I am missing something here or whether this might actually be a bug in the API: https://github.com/cefsharp/CefSharp/discussions/3922. I hope we'll get a response quickly so that I can fix the issue.

**Comment by dbuechel** [comment]
We were able to solve the issue with redirects thanks to the help of the CefSharp maintainer. The root cause actually does appear to be (yet another) bug in the browser engine, which we luckily were able to circumvent by using another API.

@ScottDellinger You can retest the issue using the latest development build if you'd like: https://sebdev-let.ethz.ch/api/buildjobs/t4047847xq50t0v5/artifacts/SEB_3.3.2.397_SetupBundle.exe

**Comment by jeroenhabets** [comment]
@dbuechel Great news! I will notify my team so they can test as well.

When do you/SEB plan to release 3.3.2? (The Github 3.3.2 project shows this ticket In Progress and the rest in To Do. Do all those items have to be done before a 3.3.2 release?)

**Comment by dbuechel** [comment]
@jeroenhabets Excellent, that's great!

> When do you/SEB plan to release 3.3.2? (The Github 3.3.2 project shows this ticket In Progress and the rest in To Do. Do all those items have to be done before a 3.3.2 release?)

Yes, we originally planned the release for end of Q4 2021. But due to unforeseeable events, we won't be able to keep up that deadline. We'll re-prioritize and hopefully be able to release version 3.3.2 (with potentially a few less issues fixed) by the end of January 2022.

**Comment by jeroenhabets** [comment]
@dbuechel we spotted one issue between our MacOS and Windows tests, logged as [#310](https://github.com/SafeExamBrowser/seb-win-refactoring/issues/310)

**Comment by jeroenhabets** [comment]
@dbuechel, @danschlet our test with [SEB_3.3.2.397_SetupBundle.exe](https://github.com/SafeExamBrowser/seb-win-refactoring/issues/84#issuecomment-997813289) was successful 🚀 🍾 !
@danschlet But the same `.seb` with [SEB Mac 3.2pre5](https://github.com/SafeExamBrowser/seb-mac/releases/tag/3.2pre5) with "browserWindowWebView to the policy "Prefer Modern" (integer value 3)" our validation failed.

1) Is that the correct SEB Mac version to test with? (Or did you need to fix some redirect/URL calculations as well?)
2) If so, how and where do you want me to report my findings (here in this thread or new seb-mac issue)? And if you would want access to our test system how to provide it privately?

**Comment by danschlet** [comment]
Please create an issue in seb-mac and mention its number here.

As GitHub still doesn't support private messages, this is annoying. You could write a pn on our old SourceForge site or use info at safeexambrowser . org (but please only for access data, we don't give support by e-mail to non-SEB Alliance members).

**Comment by jeroenhabets** [comment]
@danschlet OK, thank you very much! I've logged [seb-mac #174](https://github.com/SafeExamBrowser/seb-mac/issues/174) and send you a mail with @dbuechel in CC.

Both your work is dearly appreciated! Happy New Year 🚀 🎉 !

**Comment by dbuechel** [comment]
@jeroenhabets Thanks a lot, same to you as well! And many thanks for the extensive testing and help to all involved, it really is highly appreciated and allows us to solve issues more efficiently and thus deliver new versions faster.

**Comment by jeroenhabets** [comment]
FYI: Closed my [SEB Mac issue 174](https://github.com/SafeExamBrowser/seb-mac/issues/174) as I can no longer reproduce it. 
Our Javascript API key check is now working fine with the same SEB file on both Mac OS Monterey and Win10 (in VirtualBox) 🎉 

Thanks for your help! Much appreciated!

I don't know if I am supposed to Close this ticket (?), but I also consider it **Done** and we're only waiting for the 3.3.2 release now. @dbuechel do you have a date when you plan this to be released?

**Comment by dbuechel** [comment]
Excellent, thanks for the update @jeroenhabets. I'll normally close the tickets once the release has been done, but this doesn't update the milestone progress correctly (even though the issue is set to done), so I might change this in the future. Yes, we plan to release version 3.3.2 no later than by the end of January, if all is going according to plan.

**Comment by danschlet** [comment]
I'm happy that it now also works on macOS for you, we did some testing with the new Moodle implementation and didn't observe any issues there with Mac/Windows, so I was confused about your bug report...

**Comment by dbuechel** [comment]
Now released with version 3.3.2.
